### ISSUE TITLE
ConcurrentModificationException when a target depends on itself

### ISSUE THREAD

**Issue by pvcnt** [pantsbuild/intellij-pants-plugin#191]
Hi,

I got a ConcurrentModificationException when refreshing a Pants project where, after a refactoring, a target ended up depending on itself. Maybe we should get a better looking error message. :-)

Thanks.

```
Error:exception during working with external system: java.util.ConcurrentModificationException
    at java.util.TreeMap$PrivateEntryIterator.nextEntry(TreeMap.java:1211)
    at java.util.TreeMap$KeyIterator.next(TreeMap.java:1265)
    at com.twitter.intellij.pants.service.project.modifier.PantsCyclicDependenciesModifier.modify(PantsCyclicDependenciesModifier.java:32)
    at com.twitter.intellij.pants.service.project.PantsResolver.addInfoTo(PantsResolver.java:83)
    at com.twitter.intellij.pants.service.project.PantsSystemProjectResolver.resolveUsingPantsGoal(PantsSystemProjectResolver.java:199)
    at com.twitter.intellij.pants.service.project.PantsSystemProjectResolver.resolveProjectInfoImpl(PantsSystemProjectResolver.java:139)
    at com.twitter.intellij.pants.service.project.PantsSystemProjectResolver.resolveProjectInfo(PantsSystemProjectResolver.java:75)
    at com.twitter.intellij.pants.service.project.PantsSystemProjectResolver.resolveProjectInfo(PantsSystemProjectResolver.java:53)
    at com.intellij.openapi.externalSystem.service.remote.RemoteExternalSystemProjectResolverImpl.a(RemoteExternalSystemProjectResolverImpl.java:37)
    at com.intellij.openapi.externalSystem.service.remote.AbstractRemoteExternalSystemService.execute(AbstractRemoteExternalSystemService.java:59)
    at com.intellij.openapi.externalSystem.service.remote.RemoteExternalSystemProjectResolverImpl.resolveProjectInfo(RemoteExternalSystemProjectResolverImpl.java:37)
    at com.intellij.openapi.externalSystem.service.remote.wrapper.ExternalSystemProjectResolverWrapper.resolveProjectInfo(ExternalSystemProjectResolverWrapper.java:49)
    at com.intellij.openapi.externalSystem.service.internal.ExternalSystemResolveProjectTask.doExecute(ExternalSystemResolveProjectTask.java:51)
    at com.intellij.openapi.externalSystem.service.internal.AbstractExternalSystemTask.execute(AbstractExternalSystemTask.java:138)
    at com.intellij.openapi.externalSystem.service.internal.AbstractExternalSystemTask.execute(AbstractExternalSystemTask.java:124)
    at com.intellij.openapi.externalSystem.util.ExternalSystemUtil$3.execute(ExternalSystemUtil.java:415)
    at com.intellij.openapi.externalSystem.util.ExternalSystemUtil$5.run(ExternalSystemUtil.java:494)
    at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:635)
    at com.intellij.openapi.progress.impl.CoreProgressManager$3.run(CoreProgressManager.java:170)
    at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:494)
    at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:443)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:54)
    at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:155)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl$1.run(ProgressManagerImpl.java:128)
    at com.intellij.openapi.application.impl.ApplicationImpl$2.run(ApplicationImpl.java:308)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by wisechengyi** [comment]
What does pants export <target> say? I'd imagine export will fail if target depends on itself, in which case IntelliJ will throw a more understandable message.

I'd guess this is caused by something else. 

Also make sure the plugin is 1.4.3 or up.

**Comment by pvcnt** [comment]
Sorry for the delay, I successfully reproduced the bug with the same condition: a self-dependency. The output of pants export seems correct (i.e., it completed with no error) and does include the self-dependency under the `targets` key associated with my target.

**Comment by wisechengyi** [comment]
I did the following in with pantsbuilds/pants example then imported the project, but it does repro.
```
diff --git a/examples/tests/java/org/pantsbuild/example/hello/greet/BUILD b/examples/tests/java/org/pantsbuild/example/hello/greet/BUILD
index 5a9cda3..e26d456 100644
--- a/examples/tests/java/org/pantsbuild/example/hello/greet/BUILD
+++ b/examples/tests/java/org/pantsbuild/example/hello/greet/BUILD
@@ -8,6 +8,7 @@ junit_tests(name='greet',
   dependencies=[
     '3rdparty:junit',
     'examples/src/java/org/pantsbuild/example/hello/greet',
+    'examples/tests/java/org/pantsbuild/example/hello/greet:greet'
   ],
   resources=[
     'examples/src/resources/org/pantsbuild/example/hello',
```

**Comment by wisechengyi** [comment]
implemented in #288

**Comment by wisechengyi** [comment]
available at master build https://plugins.jetbrains.com/plugin/7412-pants-support/update/34069, and will go in at the next release.
